Task_1_4_1/src/main/java/RecordBook.java

```diff
@@ -0,0 +1,153 @@
+            return 3;
+        } else if (grade.compareTo("хорошо") == 0) {
+            return 4;
+        } else if (grade.compareTo("отлично") == 0) {
+            return 5;
+        } else {
+            return 0;
+        }
+    }
+
+    /**
+     * Function for putting the grade into the record book.
+     * If the grade for a particular subject in the semester has already been set, then it can be changed.
+     *
+     * @param semester semester number
+     * @param subject name of the subject
+     * @param mark mark for this subject (valid: "удовлетворительно", "хорошо", "отлично")
+     */
+    public void putGrade(int semester, String subject, String mark) throws IllegalArgumentException {
+        int grade = convertGrade(mark);
+        if (grade == 0) {
+            throw new IllegalArgumentException("Invalid mark");
+        }
+
+        HashMap<String, Integer> temp = this.grades.get(semester);
+        if (temp.containsKey(subject)) { //changing an existing mark
+            this.amountOfAllGrades--;
+            this.sumOfAllGrades -= temp.get(subject);
+            if (temp.get(subject) == 3) {
+                this.amountOfThrees--;
+            }
+        }
+        temp.put(subject, grade);
+        this.grades.put(semester, temp);
+
+        if (grade == 3) {
+            this.amountOfThrees++;
+        }
+
+        this.amountOfAllGrades++;
+        this.sumOfAllGrades += grade;
+
+        if (this.finalGrades.containsKey(subject)) {
+            ArrayList<Integer> tmp = this.finalGrades.get(subject);
+            if (tmp.get(0) < semester) {
+                tmp.add(0, semester);
+                tmp.add(1, grade);
+            } else if ((tmp.get(0) == semester) && (tmp.get(1) != grade)) {
+                tmp.add(1, grade);
+            }
+        } else {
+            ArrayList<Integer> tmp = new ArrayList<>();
+            tmp.add(semester);
+            tmp.add(grade);
+            this.finalGrades.put(subject, tmp);
+        }
+    }
+
+    /**
+     * Function that returns the current average score for the all studying time.
+     *
+     * @return current average score
+     */
+    public Double giveCurrentAverageScore() {
+        return (double) this.sumOfAllGrades / (double) this.amountOfAllGrades;
+    }
+
+    /**
+     * Function that returns the possibility of getting an increased scholarship in the specified
+     * semester. An increased scholarship is given if all grades in the current semester are 5.
+     *
+     * @param semester current semester
+     * @return true - if the increased scholarship is given; false - if the increased scholarship
+     * is not given
+     */
+    public Boolean givePossibilityGetIncreasedScholarship(int semester) {
+        HashMap<String, Integer> temp = this.grades.get(semester);
+        for (Map.Entry<String, Integer> entry : temp.entrySet()) {
+            if (entry.getValue() != 5) {
+                return false;
+            }
+        }
+        return true;
```

Comment:
Здесь можно сделать проще через стримы
```suggestion
     return temp.values().stream().allMatch(grade -> grade == 5);

```